### PR TITLE
#199 Markdown specification (part 2 of 2)

### PR DESCRIPTION
**Pull request by kalbfled**
# Description

These changes add more files to the tests/test_files/ subfolders to codify the expected behavior when rendering markdown as HTML or plain text, and they implement unit tests for all the test_files/ files.  The new test files focus on personalization substitutions into links and block quotes, including the substitution of lists into the latter.

Substitutions into URLs of text containing spaces should be URL-safe encoded.  The goal isn't that the link is valid (that's on the user); it's that the visual presentation is correct.

The unit tests include a simple happy path test and a test for missing personalization fields, both using markdown for a paragraph.  I did not exhaustively test placeholders in all the various markdown permutations we might encounter because I don't think that would add value.  When the tests I actually created for the hard cases pass, I am confident the functionality will work for the more straight-forward cases.

You might note that the 3 test classes I created are repetitive.  Each class tests one input markdown file.  With additional parametrization, I could roll all of these classes into a single test, but I rejected that implementation in favor of only loading the input files once, via a class-scoped fixture.

issue #199

## How Has This Been Tested?

I ran the unit tests.  The tests I added with this PR all run and fail with NotImplementedError, as is appropriate now.  Subsequently, I skipped all these new tests.  There are no changes required in notification-api.  The next steps are #213, which builds out the implement needed to make these tests pass, and #215, which will integrate the new implementation with notification-api.

## Checklist

- [x] I have assigned myself to this PR
- [x] PR has an [appropriate title](https://github.com/department-of-veterans-affairs/vanotify-team/blob/main/Engineering/team_agreement.md#naming-prs): #9999 - What the thing does
- [x] PR has a detailed description, including links to specific documentation
- [ ] I have added the [appropriate labels](https://github.com/department-of-veterans-affairs/notification-api/blob/main/.github/release.yaml) to the PR.
- [x] I did not remove any parts of the template, such as checkboxes even if they are not used
- [x] My code follows the [style guidelines](https://github.com/department-of-veterans-affairs/vanotify-team/blob/main/Process/style_guide.md) of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to any documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works. [Testing guidelines](https://github.com/department-of-veterans-affairs/vanotify-team/blob/main/Process/testing_guide.md)
- [x] I have ensured the latest main is merged into my branch and all checks are green prior to review
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [x] The ticket was moved into the DEV test column when I began testing this change

### CODE REVIEW

File: notifications_utils/template2.py [hunk 1]
```diff
@@ -0,0 +1,39 @@
+def render_notify_markdown(markdown: str, personalization: dict = {}, as_html: bool = True) -> str:
+    """
+    Substitute personalization values into markdown, and return the markdown as HTML or plain text.
+    """
+
+    # TODO - Perform substitutions in the markdown.  Raise ValueError for missing fields.
```
Comment (MackHalliday) [inline notifications_utils/template2.py:6]: Do we need a ticket number on this TODO

File: notifications_utils/template2.py [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+    The two "content" parameters generally are the output of render_notify_markdown (above).
+
+    returns: A 2-tuple in which the first value is the full HTML e-mail; the second, the plain text e-mail.
+    """
+
+    if html_content is None and plain_text_content is None:
+        raise ValueError('You must supply one of these parameters.')
+
+    # TODO - Perform substitutions in the subject.  Raise ValueError for missing fields.
```
Comment (MackHalliday) [inline notifications_utils/template2.py:36]: More TODOs - Do we need a ticket number

File: tests/test_files/markdown/placeholders/links_placeholders.md [hunk 3]
```diff
@@ -0,0 +1,25 @@
+# Links With Placeholders
```
Comment (cris-oddball) [inline tests/test_files/markdown/placeholders/links_placeholders.md:1]: I don't see that all the use cases from the template QA All the Formatting are tested here. Please include them. The below should e enough information for you to go into Portal and edit the template to see the markdown. Below is what we get back from the templates routes for the placeholder in links section.

DEV VANotify Template ID "86820037-1374-4072-9f9f-e74f235170a1"

```html
Fixed Cases:\n1. placeholder at end of [markdown link](https://test.com/((foo)))\n2. placeholder at start of [markdown link](((foo))test.com)\n3. placeholder in middle of [markdown link](https://((foo))-test.com/)\n4. placeholder as [markdown link](((foo)))\n5. multiple placeholders in [markdown link](https://((foo))-test.com?x=((foo)))\n\nEdge Cases: (less likely to happen, but working as expected)\n1. placeholder within markdown link and markdown text [markdown ((foo)) link](https://test.com/((foo)))\n2. multiple placeholders in as query params [markdown link](https://((foo))-test.com?x=((foo))&y=((foo)))\n\n---\n\nA 

```

**Review by cris-oddball — CHANGES_REQUESTED** [review]
There may be more, but I need to switch over to testing another ticket.

File: notifications_utils/template2.py (hunk 2 above)
Comment (kalbfled) [reply notifications_utils/template2.py]: added

File: notifications_utils/template2.py [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+    """
+
+    # TODO - Perform substitutions in the markdown.  Raise ValueError for missing fields.
+
+    if as_html:
+        # TODO - pass the markdown to the HTML renderer
+        pass
+    else:
+        # TODO - pass the markdown to the plain text renderer
+        pass
```
[the remark this answers is not in the document]
Comment (kalbfled) [reply notifications_utils/template2.py]: added

File: notifications_utils/template2.py (hunk 1 above)
Comment (kalbfled) [reply notifications_utils/template2.py]: added

File: notifications_utils/template2.py [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+
+    raise NotImplementedError
+
+
+# TODO - The signature and return type might change after #213, during integration with notifcation-api.
```
[the remark this answers is not in the document]
Comment (kalbfled) [reply notifications_utils/template2.py]: added

File: notifications_utils/template2.py [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+    The two "content" parameters generally are the output of render_notify_markdown (above).
+
+    returns: A 2-tuple in which the first value is the full HTML e-mail; the second, the plain text e-mail.
+    """
+
+    if html_content is None and plain_text_content is None:
+        raise ValueError('You must supply one of these parameters.')
+
+    # TODO - Perform substitutions in the subject.  Raise ValueError for missing fields.
+    # TODO - Jinja2 template substitution
```
[the remark this answers is not in the document]
Comment (kalbfled) [reply notifications_utils/template2.py]: added

File: tests/test_files/markdown/placeholders/links_placeholders.md (hunk 3 above)
Comment (kalbfled) [reply tests/test_files/markdown/placeholders/links_placeholders.md:1]: These are in addition to what is already tested via tests/test_files/markdown/links.md (#198).  Placeholder tests focus specifically on what could go wrong during substitution.  In the case of links, it's spaces in the URL.  For block quotes, it's the insertion of lists.

File: notifications_utils/template2.py [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+    plain_text_content: str | None = None,
+    subject_personalization: dict = {}
+) -> tuple[str | None, str | None]:
+    """
+    In addition to the content body, e-mail notifications might have personalization values in the
+    subject line, and the content body might be plugged into a Jinja2 template.
+
+    The two "content" parameters generally are the output of render_notify_markdown (above).
+
+    returns: A 2-tuple in which the first value is the full HTML e-mail; the second, the plain text e-mail.
```
[the remark this answers is not in the document]
Comment (kalbfled) [reply notifications_utils/template2.py:30]: I think that's accurate for ENP, but I was not attempting to emulate that here.  It's more of a "note to self" at this point.  #215 will revisit function.

**Review by cris-oddball — APPROVED** [review]

File: tests/test_files/markdown/placeholders/links_placeholders.md (hunk 3 above)
Comment (cris-oddball) [reply tests/test_files/markdown/placeholders/links_placeholders.md:1]: Resolved question on a call on Friday

File: tests/test_template2.py [hunk 8]
```diff
@@ -0,0 +1,238 @@
+from os import scandir
+from typing import Generator
+
+import pytest
+
+from notifications_utils.template2 import render_notify_markdown
+
+
+# Skip all tests in this module.
```
Comment (cris-oddball) [inline tests/test_template2.py:9]: This should be a TODO #213 here. I won't insist on it. The benefit of TODO is that most IDEs will look for those and report them out when queried.

**Review by cris-oddball — CHANGES_REQUESTED** [review]

File: tests/test_template2.py (hunk 8 above)
Comment (kalbfled) [reply tests/test_template2.py:9]: I added a TODO.

File: notifications_utils/template2.py [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+        pass
+
+    raise NotImplementedError
+
+
+# TODO - The signature and return type might change for #215 or later, during integration with notifcation-api.
+def render_email(
+    html_content: str | None = None,
+    plain_text_content: str | None = None,
+    subject_personalization: dict = {}
```
[the remark this answers is not in the document]
Comment (kalbfled) [reply notifications_utils/template2.py]: I made None the default.

**Review by k-macmillan — APPROVED** [review]
Approved, would still appreciate an answer to the question though regarding action links.

File: tests/test_files/markdown/placeholders/action_links_placeholders.md [hunk 10]
```diff
@@ -0,0 +1,5 @@
+>>[((url_text))](https://www.youtube.com/watch?v=((yt_video_id)))
+
+>>[action link](((url)))
+
+>>[((url_text))](((url)))
```
[the remark this answers is not in the document]
Comment (cris-oddball) [reply tests/test_files/markdown/placeholders/action_links_placeholders.md:5]: That was done on the previous ticket.
https://github.com/department-of-veterans-affairs/notification-utils/blob/main/tests/test_files/markdown/block_quotes_action_link.md

[the remark this answers is not in the document]
Comment (kalbfled) [reply tests/test_files/markdown/placeholders/action_links_placeholders.md:5]: I test action links in block quotes [here](https://github.com/department-of-veterans-affairs/notification-utils/blob/main/tests/test_files/html/block_quotes_action_link.html), without placeholders.  I did not create new tests with placeholders.  If substitutions work in block quotes generally, they will work in action links.  I test substitution in block quotes elsewhere in this PR.

**Review by cris-oddball — APPROVED** [review]

**Review by MackHalliday — APPROVED** [review]
Approved - Reviewed recent commits
